register of size 1?

File: MB9AF10xN.svd

Contents:

``` xml
        <register>
          <name>WDG_RIS</name>
          <description>Hardware Watchdog Timer Interrupt Status Register</description>
          <addressOffset>0x10</addressOffset>
          <size>1</size>
          <access>read-only</access>
          <resetValue>0xFF</resetValue>
          <resetMask>0x00</resetMask>
          <!-- FIELDS -->
          <fields>
            <!-- FIELD "RIS" -->
            <field>
              <name>RIS</name>
              <description>Hardware watchdog interrupt status bit</description>
              <lsb>0</lsb>
              <msb>0</msb>
              <access>read-only</access>
            </field>
          </fields>
        </register>
```

Produces:

``` rust
        impl R {
            # [ doc = r" Value of the register as raw bits" ]
            # [ inline ( always ) ]
            pub fn bits(&self) -> bool {
                self.bits
            }
            # [ doc = "Bit 0 - Hardware watchdog interrupt status bit" ]
            # [ inline ( always ) ]
            pub fn ris(&self) -> RISR {
                let bits = {
                    const MASK: bool = true;
                    const OFFSET: u8 = 0;
                    ((self.bits >> OFFSET) & MASK as bool) != 0
                };
                RISR { bits }
            }
        }
```

Compiler error:

``` console
error[E0369]: binary operation `>>` cannot be applied to type `bool`
    --> src/lib.rs:6496:22
     |
6496 |                     ((self.bits >> OFFSET) & MASK as bool) != 0
     |                      ^^^^^^^^^^^^^^^^^^^^^
     |
     = note: an implementation of `std::ops::Shr` might be missing for `bool`

error: aborting due to previous error(s)
```

The problem I see that the SVD file declares the register has having a size of 1. I'm not sure if that allowed by the SVD format (the website appears to be down ATM) but I'd expect it to only allow sizes multiple of 8 (~~bytes~~ bits). Note that we are talking about *registers*; bitfields can have any size (smaller than the size of their register).

Anyhow the current implementation maps a size of 1 to a `bool` so the `bits` field of `R` is `bool` which can be shifted to the right by `OFFSET` thus the compiler error.

cc @whitequark fallout from #84 

EDIT I meant *bits* not *bytes*.

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -445,96 +445,96 @@ main() {
 
         Spansion)
             # OK
-            test_svd MB9AF12xK
-            test_svd MB9AF12xL
-            test_svd MB9AF42xK
-            test_svd MB9AF42xL
-            test_svd MB9BF12xJ
-            test_svd MB9BF12xS
-            test_svd MB9BF12xT
-            test_svd MB9BF16xx
-            test_svd MB9BF32xS
-            test_svd MB9BF32xT
-            test_svd MB9BF36xx
-            test_svd MB9BF42xS
-            test_svd MB9BF42xT
-            test_svd MB9BF46xx
-            test_svd MB9BF52xS
-            test_svd MB9BF52xT
-            test_svd MB9BF56xx
+            # test_svd MB9AF12xK
+            # test_svd MB9AF12xL
+            # test_svd MB9AF42xK
+            # test_svd MB9AF42xL
+            # test_svd MB9BF12xJ
+            # test_svd MB9BF12xS
+            # test_svd MB9BF12xT
+            # test_svd MB9BF16xx
+            # test_svd MB9BF32xS
+            # test_svd MB9BF32xT
+            # test_svd MB9BF36xx
+            # test_svd MB9BF42xS
+            # test_svd MB9BF42xT
+            # test_svd MB9BF46xx
+            # test_svd MB9BF52xS
+            # test_svd MB9BF52xT
+            # test_svd MB9BF56xx
 
-            # FIXME(#102) "binary operation `>>` cannot be applied to type `bool`"
-            # test_svd MB9AF10xN
-            # test_svd MB9AF10xR
-            # test_svd MB9AF11xK
-            # test_svd MB9AF11xL
-            # test_svd MB9AF11xM
-            # test_svd MB9AF11xN
-            # test_svd MB9AF13xK
-            # test_svd MB9AF13xL
-            # test_svd MB9AF13xM
-            # test_svd MB9AF13xN
-            # test_svd MB9AF14xL
-            # test_svd MB9AF14xM
-            # test_svd MB9AF14xN
-            # test_svd MB9AF15xM
-            # test_svd MB9AF15xN
-            # test_svd MB9AF15xR
-            # test_svd MB9AF31xK
-            # test_svd MB9AF31xL
-            # test_svd MB9AF31xM
-            # test_svd MB9AF31xN
-            # test_svd MB9AF34xL
-            # test_svd MB9AF34xM
-            # test_svd MB9AF34xN
-            # test_svd MB9AFA3xL
-            # test_svd MB9AFA3xM
-            # test_svd MB9AFA3xN
-            # test_svd MB9AFA4xL
-            # test_svd MB9AFA4xM
-            # test_svd MB9AFA4xN
-            # test_svd MB9AFB4xL
-            # test_svd MB9AFB4xM
-            # test_svd MB9AFB4xN
-            # test_svd MB9BF10xN
-            # test_svd MB9BF10xR
-            # test_svd MB9BF11xN
-            # test_svd MB9BF11xR
-            # test_svd MB9BF11xS
-            # test_svd MB9BF11xT
-            # test_svd MB9BF12xK
-            # test_svd MB9BF12xL
-            # test_svd MB9BF12xM
-            # test_svd MB9BF21xS
-            # test_svd MB9BF21xT
-            # test_svd MB9BF30xN
-            # test_svd MB9BF30xR
-            # test_svd MB9BF31xN
-            # test_svd MB9BF31xR
-            # test_svd MB9BF31xS
-            # test_svd MB9BF31xT
-            # test_svd MB9BF32xK
-            # test_svd MB9BF32xL
-            # test_svd MB9BF32xM
-            # test_svd MB9BF40xN
-            # test_svd MB9BF40xR
-            # test_svd MB9BF41xN
-            # test_svd MB9BF41xR
-            # test_svd MB9BF41xS
-            # test_svd MB9BF41xT
-            # test_svd MB9BF50xN
-            # test_svd MB9BF50xR
-            # test_svd MB9BF51xN
-            # test_svd MB9BF51xR
-            # test_svd MB9BF51xS
-            # test_svd MB9BF51xT
-            # test_svd MB9BF52xK
-            # test_svd MB9BF52xL
-            # test_svd MB9BF52xM
-            # test_svd MB9BF61xS
-            # test_svd MB9BF61xT
-            # test_svd MB9BFD1xS
-            # test_svd MB9BFD1xT
+            # #102 regression tests
+            test_svd MB9AF10xN
+            test_svd MB9AF10xR
+            test_svd MB9AF11xK
+            test_svd MB9AF11xL
+            test_svd MB9AF11xM
+            test_svd MB9AF11xN
+            test_svd MB9AF13xK
+            test_svd MB9AF13xL
+            test_svd MB9AF13xM
+            test_svd MB9AF13xN
+            test_svd MB9AF14xL
+            test_svd MB9AF14xM
+            test_svd MB9AF14xN
+            test_svd MB9AF15xM
+            test_svd MB9AF15xN
+            test_svd MB9AF15xR
+            test_svd MB9AF31xK
+            test_svd MB9AF31xL
+            test_svd MB9AF31xM
+            test_svd MB9AF31xN
+            test_svd MB9AF34xL
+            test_svd MB9AF34xM
+            test_svd MB9AF34xN
+            test_svd MB9AFA3xL
+            test_svd MB9AFA3xM
+            test_svd MB9AFA3xN
+            test_svd MB9AFA4xL
+            test_svd MB9AFA4xM
+            test_svd MB9AFA4xN
+            test_svd MB9AFB4xL
+            test_svd MB9AFB4xM
+            test_svd MB9AFB4xN
+            test_svd MB9BF10xN
+            test_svd MB9BF10xR
+            test_svd MB9BF11xN
+            test_svd MB9BF11xR
+            test_svd MB9BF11xS
+            test_svd MB9BF11xT
+            test_svd MB9BF12xK
+            test_svd MB9BF12xL
+            test_svd MB9BF12xM
+            test_svd MB9BF21xS
+            test_svd MB9BF21xT
+            test_svd MB9BF30xN
+            test_svd MB9BF30xR
+            test_svd MB9BF31xN
+            test_svd MB9BF31xR
+            test_svd MB9BF31xS
+            test_svd MB9BF31xT
+            test_svd MB9BF32xK
+            test_svd MB9BF32xL
+            test_svd MB9BF32xM
+            test_svd MB9BF40xN
+            test_svd MB9BF40xR
+            test_svd MB9BF41xN
+            test_svd MB9BF41xR
+            test_svd MB9BF41xS
+            test_svd MB9BF41xT
+            test_svd MB9BF50xN
+            test_svd MB9BF50xR
+            test_svd MB9BF51xN
+            test_svd MB9BF51xR
+            test_svd MB9BF51xS
+            test_svd MB9BF51xT
+            test_svd MB9BF52xK
+            test_svd MB9BF52xL
+            test_svd MB9BF52xM
+            test_svd MB9BF61xS
+            test_svd MB9BF61xT
+            test_svd MB9BFD1xS
+            test_svd MB9BFD1xT
         ;;
 
         STMicro)

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -579,7 +579,7 @@ main() {
             test_svd STM32L4x6
             test_svd STM32W108
 
-            # FIXME(???) "field is never used: `register`"
+            # FIXME(#91) "field is never used: `register`"
             # test_svd STM32L051x
             # test_svd STM32L052x
             # test_svd STM32L053x

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -130,6 +130,13 @@ main() {
             # test_svd MKV58F22
             # test_svd MKV58F24
 
+            # BAD-SVD field names are equivalent when case is ignored
+            # test_svd MK61F15
+            # test_svd MK61F15WS
+            # test_svd MK70F12
+            # test_svd MK70F15
+            # test_svd MK70F15WS
+
             # OK
             # NOTE it would take too long to test all these so we only test half
             test_svd MK02F12810
@@ -222,13 +229,6 @@ main() {
             test_svd MKW31Z4
             # test_svd MKW40Z4
             # test_svd MKW41Z4
-
-            # FIXME(#105) "error[E0428]: a type named `NOT_USEDR` has already been defined in this module"
-            # test_svd MK61F15
-            # test_svd MK61F15WS
-            # test_svd MK70F12
-            # test_svd MK70F15
-            # test_svd MK70F15WS
 
             # FIXME(#92) "duplicate definitions with name `bits`"
             # test_svd MKE02Z2

--- a/src/generate.rs
+++ b/src/generate.rs
@@ -435,6 +435,13 @@ pub fn register(
                         format!("Register {} has no `size` field",
                                 register.name)
                     })?;
+    let rsize = if rsize < 8 {
+        8
+    } else if rsize.is_power_of_two() {
+        rsize
+    } else {
+        rsize.next_power_of_two()
+    };
     let rty = rsize.to_ty()?;
     let description = util::respace(&register.description);
 

